File: csharp/1103.distribute-candies-to-people.cs
/*
 * @lc app=leetcode id=1103 lang=csharp
 *
 * [1103] Distribute Candies to People
 */

using System;

// @lc code=start
public class Solution {
    public int[] DistributeCandies(int candies, int num_people) {
        var ret = new int[num_people];
        int i = 0, acc = 1;
        while (candies > 0) {
            int c = Math.Min(acc++, candies);
            candies -= c;
            ret[i] += c;
            i = (i + 1) % num_people;
        }
        return ret;
    }
}
// @lc code=end

File: java/1103.distribute-candies-to-people.java
/*
 * @lc app=leetcode id=1103 lang=java
 *
 * [1103] Distribute Candies to People
 */

// @lc code=start
class Solution {
    public int[] distributeCandies(int candies, int numPeople) {
        int m, k;
        for (k = 0, m = 0; (1 + m + numPeople) * (m + numPeople) / 2 <= candies; m += numPeople, ++k);
        int r = candies - (1 + m) * m / 2;
        int[] ret = new int[numPeople];
        for (int i = 1; i <= numPeople; ++i) {
            int s = Math.min(r, k * numPeople + i);
            ret[i - 1] = s + k * i + (k - 1) * k / 2 * numPeople;
            r -= s;
        }
        return ret;
    }
}
// @lc code=end

